test: assumeutxo: add missing tests in wallet_assumeutxo.py

alfonsoromanz opened the pull request:
Adding tests in `./test/functional/wallet_assumeutxo.py` to cover the following scenario:
- test loading a wallet (backup) on a pruned node

alfonsoromanz commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -32,19 +28,47 @@ def add_options(self, parser):
+        self.add_nodes(3)
         self.start_nodes(extra_args=self.extra_args)
 
+    def test_descriptor_import(self, node, wallet_name, key, timestamp, should_succeed=True, expected_error_message=None):
```
Fixed. Thanks!

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -32,19 +28,47 @@ def add_options(self, parser):
         self.extra_args = [
             [],
             [],
+            ["-prune=1"]
```
You should use the `fastprune` option as well and check if there has already been some pruning happening when the test runs, i.e. check that there is some pruneheight. There isn't much of a difference if the option is set but nothing has actually been pruned yet.

alfonsoromanz replied:
Thanks for the feedback. I added the `fastprune` option to my local copy and I'm working to understand how pruning operates.

When I tried loading the wallet during background synchronization ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR149)), I noticed that there wasn't a pruneheight set yet. So, I did a manual prune using `n2.pruneblockchain(START_HEIGHT)`. After pruning, I saw a pruneheight of 300, and the test continued as expected.

However, after the background validation and sync to the tip ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR200)), the pruneheight was reset to 0. When I tried to prune the blockchain again, the pruneheight remained at 0, no matter the value I passed to pruneblockchain. To further investigate, I mined an additional 500 blocks on n2 and manually pruned again with `n2.pruneblockchain(FINAL_HEIGHT)`. This time, the pruneheight was 299.

Despite this, the wallet restoration failed with this message:

> Wallet loading failed. Prune: last wallet synchronization goes beyond pruned data. You need to -reindex (download the whole blockchain again in case of pruned node)

I understand that a pruneheight of 299 means that the node keeps blocks from 299 onwards. Since the last processed block of the wallet is also 299 and the block hash matches, I expected the restoration to work. This issue also happens with lower pruneheight values, like 242.

Is this a bug, or am I missing something?

I can confirm that I am able to retrieve the block value with height `pruneheight`. When the pruneheight was 299, I validated that block 299 was still stored in the node. The same validation was done when the pruneheight was 242, confirming block 242 was also stored. So I am not sure why the wallet restoration fails

fjahr replied:
> Thanks for the feedback. I added the fastprune option to my local copy and I'm working to understand how pruning operates.

In general, check `feature_index_prune.py`, there extensive tests using pruning and I think you can learn a bit more from that how much you have to mine for the fast pruning to work as you need it.

> However, after the background validation and sync to the tip ([here](https://github.com/bitcoin/bitcoin/pull/30455/files#diff-f763a7cf53e7a5183f66f2634baef84245919265a09d82b03dc435adea8b8faeR200)), the pruneheight was reset to 0. When I tried to prune the blockchain again, the pruneheight remained at 0, no matter the value I passed to pruneblockchain. To further investigate, I mined an additional 500 blocks on n2 and manually pruned again with n2.pruneblockchain(FINAL_HEIGHT). This time, the pruneheight was 299.

Pruning can feel weird when you operate with such low numbers, the answer is probably to play around with the number of mined blocks and the test mentioned above should help you with that. In general you have to understand that pruning means we delete db we no longer need (and fastprune makes them very small so we don't need them at a much faster rate). We no longer need them when nothing is in there that we need, i.e. no blocks that we might require are in there. Blocks are appended to these files as they come in so during normal IBD (in theory at least) we can always delete the db files in order they are created (FIFO) because the blocks come in order. But with assumeutxo the blocks don't come in order anymore. So that alone is weird and you need to keep that in mind. For the same reason `getblockfrompeer` can be hard to grasp, https://github.com/bitcoin/bitcoin/pull/23813 might be an interesting read about that. I think this is the reason you don't see results from `pruneblockchain` because there are no db files to be found that don't contain blocks we don't want to keep anymore. Mining more blocks in the test overall should fix this though, as you have seen already.

> Is this a bug, or am I missing something?

Hard to say for me with just what you have written. Can you push your current status to a branch somewhere and post the link, then I can take a look and play around with it.

alfonsoromanz replied:
Thank you very much for your explanation and suggestions! I'll definitely check `feature_index_prune.py` and https://github.com/bitcoin/bitcoin/pull/23813 to better understand how this works.

Here is a branch where I pushed my current status: [debugging_assumeutxo_tests](https://github.com/alfonsoromanz/bitcoin/tree/debugging_assumeutxo_tests)

fjahr replied:
Sorry for the long wait, I finally got around to look at your branch @alfonsoromanz . You are looking at this `wallet_last_processed_block` but that is coming out of node 0. When you load the wallet into node 2 it doesn't know anything about it, so this can be ignored. What you might be thinking of instead is the birthdate of the wallet, that's part of the backup afaict so that should be relevant. But the birthdate is earlier than 299 because the wallet seems to be created before a bunch of blocks are generated. So the error seems to make sense to me and you might want to create another wallet with a later birthdate which then should be successfully load at the place where you currently get the error with the older wallet.

fjahr replied:
Ok, I see what you mean now. After chasing down some wrong leads I think I got it. The `rescan_height` is coming from the block locator stored on disk. This is a different value than `last_processed_block`. I found that there is not guaranteed that the latest block is written when a backup is created. Since the full database is included in the backup, there is indeed something there that would cause the wallet to only need to rescan after 299. So the correct behavior IMO would be that your case 2. shouldn't fail. I confirmed this with following change: #30678. I am suggesting there that this is changed but I am not sure this behavior is really so bad in practice that it will get merged. But at least I hope that answers your question.

alfonsoromanz replied:
Thanks a lot @fjahr! Do you think that restoring a backup created in height `399` (scenario 3) would be enough for the purpose of this test?

fjahr replied:
Yeah, I think that's enough for now. I am not convinced the other cases would cover additional interesting code paths but if I'm wrong the test can always be expanded later.

alfonsoromanz replied:
Done. Thanks!

fjahr reviewed (commented):
Concept ACK

alfonsoromanz commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -4,19 +4,15 @@
 )
 from test_framework.wallet import MiniWallet
+from test_framework.wallet_util import get_generate_key
+from test_framework.descriptors import descsum_create
```
Thanks. Will fix it in my next push

alfonsoromanz replied:
Fixed

alfonsoromanz commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -32,19 +28,47 @@ def add_options(self, parser):
+
     def run_test(self):
         """
         Bring up two (disconnected) nodes, mine some new blocks on the first,
```
Thanks. I will address this in my next push along with the other feedback from fjahr

alfonsoromanz replied:
Fixed

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -45,6 +46,19 @@ def setup_network(self):
+        wrpc = node.get_wallet_rpc(wallet_name)
+        result = wrpc.importdescriptors(import_request)
+
+        if expected_error_message is None:
```
You can simplify this by just returning the result from the function and then doing these checks respectively in the two places where the function is used now. Should drop the `test_` from the name as well then.

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -185,6 +192,14 @@ def run_test(self):
+        self.connect_nodes(0, 2)
+        self.sync_blocks(nodes=(n0, n2))
+        self.wait_until(lambda: len(n2.getchainstates()['chainstates']) == 1)
+        self.log.info("Ensuring wallet can be restored from backup (pruned node)")
```
nit: You could move this to the top of this block like with the other cases (unless you duplicate anyway)

alfonsoromanz replied:
Done

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -185,6 +192,14 @@ def run_test(self):
+        # and try restoring the wallet
+        self.connect_nodes(0, 2)
+        self.sync_blocks(nodes=(n0, n2))
+        self.wait_until(lambda: len(n2.getchainstates()['chainstates']) == 1)
```
nit: The three lines are similar to above for n1, so could be put into a function together

fjahr commented:
CI seems to have an issue with the import descriptor test

```
 test  2024-08-22T15:25:46.161000Z TestFramework (ERROR): Unexpected exception caught during testing 
                                   Traceback (most recent call last):
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/wallet_assumeutxo.py", line 167, in run_test
                                       result = self.import_descriptor(n1, wallet_name, key, timestamp)
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/wallet_assumeutxo.py", line 51, in import_descriptor
                                       return wrpc.importdescriptors(import_request)
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/coverage.py", line 50, in __call__
                                       return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 127, in __call__
                                       response, status = self._request('POST', self.__url.path, postdata.encode('utf-8'))
                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 106, in _request
                                       return self._get_response()
                                              ^^^^^^^^^^^^^^^^^^^^
                                     File "/ci_container_base/ci/scratch/build/bitcoin-i686-pc-linux-gnu/test/functional/test_framework/authproxy.py", line 164, in _get_response
                                       http_response = self.__conn.getresponse()
                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^
                                     File "/usr/lib/python3.12/http/client.py", line 1428, in getresponse
                                       response.begin()
                                     File "/usr/lib/python3.12/http/client.py", line 331, in begin
                                       version, status, reason = self._read_status()
                                                                 ^^^^^^^^^^^^^^^^^^^
                                     File "/usr/lib/python3.12/http/client.py", line 300, in _read_status
                                       raise RemoteDisconnected("Remote end closed connection without"
                                   http.client.RemoteDisconnected: Remote end closed connection without response
```

alfonsoromanz commented:
Sorry for not providing an update earlier. But yes, I continue working on this. I am trying to debug the CI issues with no success so far. 

I mounted a virtual machine with ubuntu 22.04 (8GB ram), and after dealing with some libraries and memory issues I was able to run the CI job [as described here ](https://github.com/bitcoin/bitcoin/blob/master/ci/README.md).

However the jobs takes around 4-5 hours to run in the VM and I was not able to reproduce the issue yet. I was using the root user to run the command.

I am running it on a Macbook Pro with M2 Pro chip and 16GB of RAM.

Any recommendations on how to test this in my mac? I was using Parallels Desktop to virtualize but my license expired so I will move on to another alternative like VirtualBox.

alfonsoromanz commented:
> Ugh, the error is https://cirrus-ci.com/task/5110045812195328?logs=ci#L2535
> 
> ```
>  node1 stderr ./src/validation.cpp:5569 GuessVerificationProgress: Assertion `pindex->m_chain_tx_count > 0' failed. `
> ```
> 
> This means that this is a real issue that should be fixed in the real code. This test is just surfacing it.

Thanks for pointing this out. I hadn't noticed the assertion was failing.

It seems the assertion was [introduced here](https://github.com/bitcoin/bitcoin/pull/29656/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R5556), and my branch doesn't include it.

I think I'll need to rebase my local branch to see if I can reproduce the issue.

fjahr commented:
> It seems the assertion was [introduced here](https://github.com/bitcoin/bitcoin/pull/29656/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R5556), and my branch doesn't include it.

No, that's just a commit where the variable is renamed. The `Assume` was introduced here: https://github.com/bitcoin/bitcoin/pull/29370/commits/0fd915ee6bef63bb360ccc5c039a3c11676c38e3. The [assumption (pun intended)](https://github.com/bitcoin/bitcoin/pull/29370#discussion_r1479427801) was that `GuessVerificationProgress` wouldn't be called in the context of assumeutxo but that was obviously not right.

I think we'll want to do something like this for now: https://github.com/fjahr/bitcoin/commit/97a7261c8afbe4a488e6ec185cd9926e278ac3c9 Every other fix I could think of seems much more invasive and I don't think we need a perfect solution for this edge case right now.

I am thinking if this should still be fixed in v28, so that users don't see this scary `Internal bug detected` log, what do you think @maflcko ?

fjahr commented:
I have opened a separate PR #30909 with my suggested fix and your test commit cherry-picked on top of it @alfonsoromanz .

alfonsoromanz commented:
> I have opened a separate PR #30909 with my suggested fix and your test commit cherry-picked on top of it @alfonsoromanz .

Thanks a lot @fjahr !

Should I modify this PR now to be on top of yours, including only the second test? Or would it be better to wait until your PR gets merged and then modify mine?

fjahr commented:
> Should I modify this PR now to be on top of yours, including only the second test? Or would it be better to wait until your PR gets merged and then modify mine?

I would recommend you wait a bit to see what the feedback on that PR is.

fjahr commented:
@alfonsoromanz While https://github.com/bitcoin/bitcoin/pull/30909 is stalling a bit it seems like it is the way we'll move forward when there is some more review since there was no push back on concept/approach. So I would suggest you proceed here assuming that #30909 is acceptable. Ideally you would change this PR to have the second test standalone, if that's not possible you can base it on #30909.

alfonsoromanz commented:
> @alfonsoromanz While #30909 is stalling a bit it seems like it is the way we'll move forward when there is some more review since there was no push back on concept/approach. So I would suggest you proceed here assuming that #30909 is acceptable. Ideally you would change this PR to have the second test standalone, if that's not possible you can base it on #30909.

Thanks for the feedback @fjahr! I will be working on this

alfonsoromanz commented:
Pushed b1affc74c6c0cd65799e065da5744b02761ef738 to address the latest feedback provided by @fjahr. Also updated the docs as mentioned here https://github.com/bitcoin/bitcoin/pull/31556#discussion_r1899762694

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -204,6 +223,20 @@ def run_test(self):
+        assert_equal(n3.getblockchaininfo()['pruneheight'], 299)
+        error_message = "Wallet loading failed. Prune: last wallet synchronisation goes beyond pruned data. You need to -reindex (download the whole blockchain again in case of pruned node)"
+        # This backup (backup_w2.dat) was created at height 199, so it can't be restored in a node with a pruneheight of 299
+        assert_raises_rpc_error(-4,  error_message, n3.restorewallet, "w2", "backup_w2.dat")
```
nit: looks like there is a double space before `error_message`

alfonsoromanz replied:
Fixed

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -204,6 +223,20 @@ def run_test(self):
+        # After background sync, pruneheight is reset to 0, so mine 500 blocks
+        # and prune the chain again
+        self.generate(n3, nblocks=500, sync_fun=self.no_op)
+        n3.pruneblockchain(FINAL_HEIGHT)
+        assert_equal(n3.getblockchaininfo()['pruneheight'], 299)
```
nit: `pruneblockchain` returns the new pruneheight, so I think you could shorten this to

``` suggestion
        assert_equal(n3.pruneblockchain(FINAL_HEIGHT), 299)
```

alfonsoromanz replied:
Done. Actually I changed it to `assert_equal(n3.pruneblockchain(FINAL_HEIGHT), 298)`. This is because the `prubeblockchain` method returns the height of the last block pruned (298 in this case) while `pruneheight` from `getblockchaininfo` returns the oldest block stored in disk (299 in our case). I added a comment to clarify this.

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -35,31 +35,57 @@ def add_options(self, parser):
+
+        assert_equal(node.getblockchaininfo()["blocks"], SNAPSHOT_BASE_HEIGHT)
+
+    def wait_for_background_validation(self, node):
```
nit: I would call it `complete_background_validation` or something similar. `wait_for_*` implies that the sync was started before somewhere else, but the sync is part of the function.

alfonsoromanz replied:
This is a good point. Done

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -136,26 +162,28 @@ def run_test(self):
+
+        self.log.info("Backup from the snapshot height can be loaded during background sync (pruned node)")
+        loaded = n3.loadtxoutset(dump_output['path'])
+        n3.pruneblockchain(START_HEIGHT)
+        assert_greater_than(n3.getblockchaininfo()['pruneheight'], 0)
```
nit: similar to below

``` suggestion
        assert_greater_than(n3.pruneblockchain(START_HEIGHT), 0)
```

alfonsoromanz replied:
Done

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -204,6 +223,20 @@ def run_test(self):
+
+        self.log.info("Ensuring wallet can be restored from a backup that was created at the pruneheight (pruned node)")
+        # This backup (backup_w.dat) was created at height 299, so it can be restored in a node with a pruneheight of 299
+        n3.restorewallet("w2", "backup_w.dat")
```
nit: Would name that wallet here `w_alt` or so, just because w2 is sort of reserved for the w2 backup

alfonsoromanz replied:
Done

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -136,26 +162,28 @@ def run_test(self):
+        assert_equal(n3.getbalance(), 0)
+
+        self.log.info("Backup from before the snapshot height can't be loaded during background sync (pruned node)")
+        assert_raises_rpc_error(-4,  error_message, n3.restorewallet, "w2", "backup_w2.dat")
```
nit: Here also two spaces

alfonsoromanz replied:
Fixed

fjahr commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -35,31 +35,57 @@ def add_options(self, parser):
+
+        normal, snapshot = node.getchainstates()["chainstates"]
+        assert_equal(normal['blocks'], START_HEIGHT)
+        assert_equal(normal.get('snapshot_blockhash'), None)
```
nit: I would have used this, it's a little more readable and a little more strict

``` suggestion
        assert 'snapshot_blockhash' not in normal
```

alfonsoromanz replied:
Done

fjahr reviewed (commented):
Did another pass, mostly minor comments, otherwise looks pretty good. Will do a final pass when my comments have been addressed.

alfonsoromanz commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -35,31 +35,57 @@ def add_options(self, parser):
+    def wait_for_background_validation(self, node):
+        self.connect_nodes(0, node.index)
+
+        self.log.info(f"Ensuring snapshot chain syncs to tip. ({FINAL_HEIGHT})")
```
Done

alfonsoromanz commented:
> Did another pass, mostly minor comments, otherwise looks pretty good. Will do a final pass when my comments have been addressed.

Thanks for the detailed feedback, @fjahr!. I pushed 2f19e42ddd85598438d0b8b49ef93c312c45c21b with your requested changes.

fjahr commented:
Code review ACK 2f19e42ddd85598438d0b8b49ef93c312c45c21b

alfonsoromanz commented:
Rebased

fjahr commented:
Re-ACK bdec262bf661db9d46fa9b4e3fb42508751b9ea8

Only rebased since last review.

fanquake commented:
@furszy you might want to review here?

furszy reviewed (commented):
Could you decouple this test into its own isolated function?
The current monolithic approach makes it harder to review, even when the changes seem simple.
See for example how [wallet_migration.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_migration.py) is structured.

alfonsoromanz commented:
Force-pushed to address @furszy's feedback

fjahr reviewed (commented):
re-ACK 55c6a69f777ac08a14e61b98efea00e5c8f98a5f

Just moved the newly added tests into separate functions (which necessitated duplicating the error message).

theStack commented:
Concept ACK

Nit for the future: putting the move-only changes on the existing code (i.e. introduction of the `validate_snapshot_import` and `complete_background_validation` methods) into an own refactoring commit would make reviewing this a bit easier, I think.

theStack commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -36,18 +32,19 @@ def add_options(self, parser):
             [],
             [],
             [],
+            ["-fastprune", "-prune=1"]
```
nit, for smaller future diff if another node is added
```suggestion
            ["-fastprune", "-prune=1"],
```

theStack reviewed (approved):
ACK 55c6a69f777ac08a14e61b98efea00e5c8f98a5f

With two non-blocking suggestions (or three, including https://github.com/bitcoin/bitcoin/pull/30455#issuecomment-3486623443).

mzumsande reviewed (commented):
Code Review ACK 55c6a69f777ac08a14e61b98efea00e5c8f98a5f

fanquake commented on test/functional/wallet_assumeutxo.py:
```diff
@@ -57,17 +54,73 @@ def import_descriptor(self, node, wallet_name, key, timestamp):
+        # and prune the chain again
+        self.generate(n3, nblocks=500, sync_fun=self.no_op)
+        assert_equal(n3.pruneblockchain(FINAL_HEIGHT), 298)  # 298 is the height of the last block pruned (pruneheight 299)
+        error_message = "Wallet loading failed. Prune: last wallet synchronisation goes beyond pruned data. You need to -reindex (download the whole blockchain again in case of pruned node)"
```
> or was the CI already red yesterday and I missed it?

The CI was green yesterday. The open/close today kicked the CI.

fanquake commented:
@alfonsoromanz can you rebase/ followup with the test failures here?

alfonsoromanz commented:
> @alfonsoromanz can you rebase/ followup with the test failures here?

Yes, I will take care of it

alfonsoromanz committed:
test, refactor snapshot import and background validation

Move snapshot import validation and background validation logic
into dedicated helper functions.

alfonsoromanz committed:
test, assumeutxo: loading a wallet (backup) on a pruned node

theStack reviewed (approved):
re-ACK cb7d5bfe4a59d6034e3a77486cbd423e1a3bfa44
(as per `$ git range-diff 55c6a69f...cb7d5bfe`)

Thanks for also tackling the nits.
